scalameta/src/main/scala/scala/meta/Origin.scala

```diff
@@ -35,6 +39,13 @@ object Origin {
     private implicit lazy val thisDialect: Dialect = this.dialect
     val start = input.tokens.apply(startTokenPos).start
     val end = if (endTokenPos != -1) input.tokens.apply(endTokenPos).end else -1
+    private def findLine(x: Int, default: Int) = input match {
+      case Input.None => default
+      case els => els.content.take(x + 1).count(_ == '\n')
+    }
+    val startLine = findLine(start, 0)
+    val endLine = findLine(end, -1)
```

Comment:
This looks like `O(n)` every time you call it where `n` is number of preceding tokens. And for all trees it's going to be `O(n!)`.